// @hallysonh/ngx-imageviewer/esm5/lib/imageviewer.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ImageViewerComponent } from './imageviewer.component';
import { IMAGEVIEWER_CONFIG, IMAGEVIEWER_CONFIG_DEFAULT } from './imageviewer.config';
var ɵ0 = IMAGEVIEWER_CONFIG_DEFAULT;
var ImageViewerModule = /** @class */ (function () {
    function ImageViewerModule() {
    }
    ImageViewerModule.decorators = [
        { type: NgModule, args: [{
                    providers: [{ provide: IMAGEVIEWER_CONFIG, useValue: ɵ0 }],
                    declarations: [ImageViewerComponent],
                    exports: [ImageViewerComponent],
                },] },
    ];
    return ImageViewerModule;
}());
export { ImageViewerModule };
function ImageViewerModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ImageViewerModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ImageViewerModule.ctorParameters;
}
export { ɵ0 };
