// @hallysonh/ngx-imageviewer/esm5/lib/imageviewer.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { InjectionToken } from '@angular/core';
var ImageViewerConfig = /** @class */ (function () {
    function ImageViewerConfig() {
    }
    return ImageViewerConfig;
}());
export { ImageViewerConfig };
function ImageViewerConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    ImageViewerConfig.prototype.width;
    /** @type {?} */
    ImageViewerConfig.prototype.height;
    /** @type {?} */
    ImageViewerConfig.prototype.bgStyle;
    /** @type {?} */
    ImageViewerConfig.prototype.scaleStep;
    /** @type {?} */
    ImageViewerConfig.prototype.rotateStepper;
    /** @type {?} */
    ImageViewerConfig.prototype.buttonStyle;
    /** @type {?} */
    ImageViewerConfig.prototype.loadingMessage;
    /** @type {?} */
    ImageViewerConfig.prototype.tooltips;
    /** @type {?} */
    ImageViewerConfig.prototype.nextPageButton;
    /** @type {?} */
    ImageViewerConfig.prototype.beforePageButton;
    /** @type {?} */
    ImageViewerConfig.prototype.zoomOutButton;
    /** @type {?} */
    ImageViewerConfig.prototype.zoomInButton;
    /** @type {?} */
    ImageViewerConfig.prototype.rotateLeftButton;
    /** @type {?} */
    ImageViewerConfig.prototype.rotateRightButton;
    /** @type {?} */
    ImageViewerConfig.prototype.resetButton;
}
/**
 * @record
 */
export function ButtonStyle() { }
function ButtonStyle_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ButtonStyle.prototype.iconFontFamily;
    /** @type {?|undefined} */
    ButtonStyle.prototype.alpha;
    /** @type {?|undefined} */
    ButtonStyle.prototype.hoverAlpha;
    /** @type {?|undefined} */
    ButtonStyle.prototype.bgStyle;
    /** @type {?|undefined} */
    ButtonStyle.prototype.iconStyle;
    /** @type {?|undefined} */
    ButtonStyle.prototype.borderStyle;
    /** @type {?|undefined} */
    ButtonStyle.prototype.borderWidth;
}
/**
 * @record
 */
export function ButtonConfig() { }
function ButtonConfig_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ButtonConfig.prototype.icon;
    /** @type {?|undefined} */
    ButtonConfig.prototype.tooltip;
    /** @type {?|undefined} */
    ButtonConfig.prototype.sortId;
    /** @type {?|undefined} */
    ButtonConfig.prototype.show;
}
/**
 * @param {?=} icon
 * @param {?=} tooltip
 * @param {?=} sortId
 * @param {?=} show
 * @return {?}
 */
export function createButtonConfig(icon, tooltip, sortId, show) {
    if (sortId === void 0) { sortId = 0; }
    if (show === void 0) { show = true; }
    return { icon: icon, tooltip: tooltip, sortId: sortId, show: show };
}
export var /** @type {?} */ IMAGEVIEWER_CONFIG = new InjectionToken('imageviewer.config');
export var /** @type {?} */ IMAGEVIEWER_CONFIG_DEFAULT = {
    width: 800,
    // component default width
    height: 600,
    // component default height
    bgStyle: '#ECEFF1',
    // component background style
    scaleStep: 0.1,
    // zoom scale step (using the zoom in/out buttons)
    rotateStepper: false,
    loadingMessage: 'Loading...',
    buttonStyle: {
        iconFontFamily: 'Material Icons',
        // font used to render the button icons
        alpha: 0.5,
        // buttons' transparence value
        hoverAlpha: 0.7,
        // buttons' transparence value when mouse is over
        bgStyle: '#000000',
        //  buttons' background style
        iconStyle: '#ffffff',
        // buttons' icon colors
        borderStyle: '#000000',
        // buttons' border style
        borderWidth: 0 // buttons' border width (0 == disabled)
    },
    tooltips: {
        enabled: true,
        // enable or disable tooltips for buttons
        bgStyle: '#000000',
        // tooltip background style
        bgAlpha: 0.5,
        // tooltip background transparence
        textStyle: '#ffffff',
        // tooltip's text style
        textAlpha: 0.9,
        // tooltip's text transparence
        padding: 15,
        // tooltip padding
        radius: 20 // tooltip border radius
    },
    nextPageButton: createButtonConfig(String.fromCharCode(0xE409), 'Next page', 0),
    beforePageButton: createButtonConfig(String.fromCharCode(0xE408), 'Previous page', 1),
    zoomOutButton: createButtonConfig(String.fromCharCode(0xE900), 'Zoom out', 0),
    zoomInButton: createButtonConfig(String.fromCharCode(0xE8FF), 'Zoom in', 1),
    rotateLeftButton: createButtonConfig(String.fromCharCode(0xE419), 'Rotate left', 2),
    rotateRightButton: createButtonConfig(String.fromCharCode(0xE41A), 'Rotate right', 3),
    resetButton: createButtonConfig(String.fromCharCode(0xE863), 'Reset', 4)
};
