// @hallysonh/ngx-imageviewer/esm5/lib/pdf.loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ResourceLoader } from './imageviewer.model';
var PdfResourceLoader = /** @class */ (function (_super) {
    tslib_1.__extends(PdfResourceLoader, _super);
    function PdfResourceLoader(_imageCache) {
        var _this = _super.call(this) || this;
        _this._imageCache = _imageCache;
        if (typeof window !== 'undefined' && 'Worker' in window) {
            if (pdfjsLib && pdfjsLib.GlobalWorkerOptions && !pdfjsLib.GlobalWorkerOptions.workerSrc) {
                pdfjsLib.GlobalWorkerOptions.workerSrc = pdfjsWorker;
            }
        }
        _this.showItemsQuantity = true;
        return _this;
    }
    /**
     * @return {?}
     */
    PdfResourceLoader.prototype.setUp = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ vm = this;
        if (vm.loading || !vm.src) {
            return;
        }
        var /** @type {?} */ loadingTask = pdfjsLib.getDocument(vm.src);
        vm.loading = true;
        vm.currentItem = 1;
        loadingTask.then(function (pdf) {
            vm._pdf = pdf;
            vm.totalItem = pdf.numPages;
            vm.loaded = true;
            vm.loadResource();
        }, function (reason) {
            console.error(reason);
        });
    };
    /**
     * @return {?}
     */
    PdfResourceLoader.prototype.loadResource = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ vm = this;
        if (!vm.loaded) {
            vm._pendingReload = true;
            return;
        }
        vm.loaded = false;
        var /** @type {?} */ url = vm.src;
        var /** @type {?} */ page = vm.currentItem;
        vm._pdf.getPage(page).then(function (pdfPage) {
            vm._page = pdfPage;
            vm.loadImage(url, page, function () {
                vm.loaded = true;
                vm.loading = false;
                if (vm._pendingReload) {
                    vm._pendingReload = false;
                    vm.loadResource();
                }
                else {
                    vm.resourceChange.next();
                }
            });
        });
    };
    /**
     * @param {?} src
     * @param {?} page
     * @param {?} onFinish
     * @return {?}
     */
    PdfResourceLoader.prototype.loadImage = /**
     * @param {?} src
     * @param {?} page
     * @param {?} onFinish
     * @return {?}
     */
    function (src, page, onFinish) {
        var /** @type {?} */ vm = this;
        var /** @type {?} */ cacheimg = vm._imageCache.getImage(src, page);
        if (cacheimg) {
            vm._image = cacheimg;
            onFinish();
            return;
        }
        var /** @type {?} */ canvas = document.createElement('canvas');
        var /** @type {?} */ context = canvas.getContext('2d');
        var /** @type {?} */ pageVp = vm._page.getViewport(2);
        canvas.width = pageVp.width;
        canvas.height = pageVp.height;
        var /** @type {?} */ renderContext = {
            canvasContext: context,
            viewport: pageVp
        };
        var /** @type {?} */ renderTask = vm._page.render(renderContext);
        renderTask.then(function () {
            canvas.toBlob(function (blob) {
                var /** @type {?} */ img = new Image();
                img.onload = onFinish;
                img.src = URL.createObjectURL(blob);
                vm._imageCache.saveImage(src, page, img);
                vm._image = img;
            });
        });
    };
    return PdfResourceLoader;
}(ResourceLoader));
export { PdfResourceLoader };
function PdfResourceLoader_tsickle_Closure_declarations() {
    /** @type {?} */
    PdfResourceLoader.prototype._pdf;
    /** @type {?} */
    PdfResourceLoader.prototype._page;
    /** @type {?} */
    PdfResourceLoader.prototype._pendingReload;
    /** @type {?} */
    PdfResourceLoader.prototype._imageCache;
}
