// @hallysonh/ngx-imageviewer/esm2015/lib/pdf.loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ResourceLoader } from './imageviewer.model';
export class PdfResourceLoader extends ResourceLoader {
    /**
     * @param {?} _imageCache
     */
    constructor(_imageCache) {
        super();
        this._imageCache = _imageCache;
        if (typeof window !== 'undefined' && 'Worker' in window) {
            if (pdfjsLib && pdfjsLib.GlobalWorkerOptions && !pdfjsLib.GlobalWorkerOptions.workerSrc) {
                pdfjsLib.GlobalWorkerOptions.workerSrc = pdfjsWorker;
            }
        }
        this.showItemsQuantity = true;
    }
    /**
     * @return {?}
     */
    setUp() {
        const /** @type {?} */ vm = this;
        if (vm.loading || !vm.src) {
            return;
        }
        const /** @type {?} */ loadingTask = pdfjsLib.getDocument(vm.src);
        vm.loading = true;
        vm.currentItem = 1;
        loadingTask.then((pdf) => {
            vm._pdf = pdf;
            vm.totalItem = pdf.numPages;
            vm.loaded = true;
            vm.loadResource();
        }, (reason) => {
            console.error(reason);
        });
    }
    /**
     * @return {?}
     */
    loadResource() {
        const /** @type {?} */ vm = this;
        if (!vm.loaded) {
            vm._pendingReload = true;
            return;
        }
        vm.loaded = false;
        const /** @type {?} */ url = vm.src;
        const /** @type {?} */ page = vm.currentItem;
        vm._pdf.getPage(page).then((pdfPage) => {
            vm._page = pdfPage;
            vm.loadImage(url, page, () => {
                vm.loaded = true;
                vm.loading = false;
                if (vm._pendingReload) {
                    vm._pendingReload = false;
                    vm.loadResource();
                }
                else {
                    vm.resourceChange.next();
                }
            });
        });
    }
    /**
     * @param {?} src
     * @param {?} page
     * @param {?} onFinish
     * @return {?}
     */
    loadImage(src, page, onFinish) {
        const /** @type {?} */ vm = this;
        const /** @type {?} */ cacheimg = vm._imageCache.getImage(src, page);
        if (cacheimg) {
            vm._image = cacheimg;
            onFinish();
            return;
        }
        const /** @type {?} */ canvas = document.createElement('canvas');
        const /** @type {?} */ context = canvas.getContext('2d');
        const /** @type {?} */ pageVp = vm._page.getViewport(2);
        canvas.width = pageVp.width;
        canvas.height = pageVp.height;
        const /** @type {?} */ renderContext = {
            canvasContext: context,
            viewport: pageVp
        };
        const /** @type {?} */ renderTask = vm._page.render(renderContext);
        renderTask.then(function () {
            canvas.toBlob(blob => {
                const /** @type {?} */ img = new Image();
                img.onload = onFinish;
                img.src = URL.createObjectURL(blob);
                vm._imageCache.saveImage(src, page, img);
                vm._image = img;
            });
        });
    }
}
function PdfResourceLoader_tsickle_Closure_declarations() {
    /** @type {?} */
    PdfResourceLoader.prototype._pdf;
    /** @type {?} */
    PdfResourceLoader.prototype._page;
    /** @type {?} */
    PdfResourceLoader.prototype._pendingReload;
    /** @type {?} */
    PdfResourceLoader.prototype._imageCache;
}
